nifi-nar-bundles/nifi-jms-bundle/nifi-jms-processors/src/main/java/org/apache/nifi/jms/processors/ConsumeJMS.java

```diff
@@ -305,6 +326,97 @@ public void accept(final JMSResponse response) {
+
+        consumer.consumeMessageSet(destinationName, errorQueueName, durable, shared, subscriptionName, messageSelector, charset, jmsResponses -> {
+            final MessageConsumer<JMSResponse> messageConsumer = new MessageConsumer.Builder<JMSResponse>()
+                    .withFlowFileWriter((new RecordWriter.Builder<JMSResponse>()
+                            .withReaderFactory(readerFactory)
+                            .withWriterFactory(writerFactory)
+                            .withSerializer(message -> message.getMessageBody() == null ? new byte[0] : message.getMessageBody())
+                            .withOutputStrategy(outputStrategy)
+                            .withLogger(getLogger()).build()))
+                    .withAttributeSupplier(message -> mergeJmsAttributes(message.getMessageHeaders(), message.getMessageProperties()))
+                    .withEventReporter(StandardRecordReceivedEventReporter.of(destinationName))
+                    .build();
+
+            messageConsumer.consumeMessages(
+                    session,
+                    jmsResponses,
+                    new MessageConsumerCallback<>() {
+                        @Override
+                        public void onSuccess(FlowFile flowFile, List<JMSResponse> processedMessages, List<JMSResponse> failedMessages) {
+                            session.transfer(flowFile, REL_SUCCESS);
+                            session.commitAsync(
+                                    () -> withLog(() -> acknowledge(processedMessages, failedMessages)),
+                                    __ -> withLog(() -> reject(processedMessages, failedMessages))
+                            );
+                            session.adjustCounter(COUNTER_RECORDS_RECEIVED, processedMessages.size() + failedMessages.size(), false);
+                            session.adjustCounter(COUNTER_RECORDS_PROCESSED, processedMessages.size(), false);
+                        }
+
+                        @Override
+                        public void onParseFailure(FlowFile flowFile, JMSResponse message, Exception e) {
+                            final FlowFile failedMessage = createFlowFileFromMessage(session, destinationName, message);
+                            session.transfer(failedMessage, REL_PARSE_FAILURE);
+                            session.adjustCounter(COUNTER_PARSE_FAILURES, 1, false);
+                        }
+
+                        @Override
+                        public void onFailure(FlowFile flowFile, List<JMSResponse> processedMessages, List<JMSResponse> failedMessages, Exception e) {
+                            reject(processedMessages, failedMessages);
+                            // It would be nicer to call rollback and yield here, but we are rethrowing the exception to have the same error handling with processSingleMessage.
+                            throw new ProcessException(e);
+                        }
+                    }
+            );
```

Comment:
The overall complexity could be drastically decreased and multiple classes (namely `StandardRecordReceivedEventReporter`, `EventReporter`, `FlowFileWriter`, `MessageConsumer`, `RecordSupplier`, `FlowFileReader`, `MessagePublisher`) could be removed when following a more direct approach at this level, like this:

```suggestion
            RecordWriter<JMSResponse> recordWriter = new RecordWriter<>(
                    readerFactory,
                    writerFactory,
                    message -> message.getMessageBody() == null ? new byte[0] : message.getMessageBody(),
                    attributeAdditionStrategy,
                    getLogger()
            );

            MessageConsumerCallback<JMSResponse> messageConsumerCallback = new MessageConsumerCallback<>() {
                @Override
                public void onSuccess(FlowFile flowFile, List<JMSResponse> processedMessages, List<JMSResponse> failedMessages) {
                    session.getProvenanceReporter().receive(flowFile, destinationName);
                    session.transfer(flowFile, REL_SUCCESS);
                    session.commitAsync(
                            () -> withLog(() -> acknowledge(processedMessages, failedMessages)),
                            commitError -> withLog(() -> reject(processedMessages, failedMessages))
                    );
                    session.adjustCounter(COUNTER_RECORDS_RECEIVED, processedMessages.size() + failedMessages.size(), false);
                    session.adjustCounter(COUNTER_RECORDS_PROCESSED, processedMessages.size(), false);
                }

                @Override
                public void onParseFailure(FlowFile flowFile, JMSResponse message, Exception e) {
                    final FlowFile failedMessage = createFlowFileFromMessage(session, destinationName, message);
                    session.transfer(failedMessage, REL_PARSE_FAILURE);
                    session.adjustCounter(COUNTER_PARSE_FAILURES, 1, false);
                }

                @Override
                public void onFailure(FlowFile flowFile, List<JMSResponse> processedMessages, List<JMSResponse> failedMessages, Exception e) {
                    reject(processedMessages, failedMessages);
                    // It would be nicer to call rollback and yield here, but we are rethrowing the exception to have the same error handling with processSingleMessage.
                    throw new ProcessException(e);
                }
            };
            AttributeSupplier<JMSResponse> attributeSupplier = message -> mergeJmsAttributes(message.getMessageHeaders(), message.getMessageProperties());
            
            recordWriter.write(
                    session,
                    jmsResponses,
                    attributeSupplier,
                    messageConsumerCallback
            );
```

Reply:
The topmost layer served multiple purposes. It has not just wrapped the difference of how provenance events needed to be generated, but it was also designed to work with demarcated and single FlowFiles. Since we agreed they are not in scope for now, I removed the extra classes.